### PR TITLE
[GPU] Enable bf16 weights

### PR DESCRIPTION
**Pull request by michal-miotk**
### Details:
 - POC for bf16 weights enablement in gpu plugin, as it is not supported yet fully by many devices (no opencl c 3.0) - contains own implementation bf16 to float32

### Tickets:
 - 143007

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp [hunk 1]
```diff
@@ -1489,6 +1489,39 @@ JitConstants MakeTypeJitConstants(Datatype dataType, const std::string& macroNam
             type_size = "0.5f";
             is_fp = false;
             break;
+        case Datatype::BF16:
+            std::cout << " macroName " << macroName << std::endl;
+            if ( macroName == "INPUT0" || macroName == "INPUT1" || macroName == "INPUT2" || macroName == "INPUT_REORDER" ) {
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/jitter.cpp:1494]: Why do we need this hack and other `macroName == "INPUT_REORDER"`?

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp [hunk 2]
```diff
@@ -1489,6 +1489,39 @@ JitConstants MakeTypeJitConstants(Datatype dataType, const std::string& macroNam
...
+                type = "ushort";
+                max_val = "USHORT_MAX";
+                min_val = "0";
+                max_func = "fmax";
+                min_func = "fmin";
+                abs_func = "fabs";
+                type_size = "2";
+                is_fp = false;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/kernel_selector/jitter.cpp:1502]: Some of these properties are not really applicable to bf16 type
Let's configure only known properties

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp (hunk 1 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/jitter.cpp]: deleted

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/include/batch_headers/common.cl [hunk 3]
```diff
@@ -61,3 +61,22 @@
 #else
 #define REQD_SUB_GROUP_SIZE(sg_size)
 #endif
+
+float convert_bf16(const ushort in){
```
[the remark this answers is not in the document]
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/include/batch_headers/common.cl]: done

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+                                                    bool,                     // reshape on decompression constants
+                                                    bool>;                    // per-tensor zero-point
+
+class BF16WeightsDecompression : public testing::WithParamInterface<MatmulWeightsDecompressionParams>,
+                                   virtual public ov::test::SubgraphBaseTest {
+public:
+    ov::CompiledModel cM;
+    ov::CompiledModel cMF16;
+    std::shared_ptr<ov::Model> functionF16;
+    ov::ParameterVector paramsf16{std::make_shared<ov::op::v0::Parameter>(ov::element::bf16, ov::Shape{{2, 2}})};
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:48]: What is the target scenario? Shouldn't bf16 data type be applied to constants, not params?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data_bf16.cl [hunk 5]
```diff
@@ -0,0 +1,219 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "include/reshape_dims.cl"
+#include "include/fetch_utils.cl"
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data_bf16.cl:6]: Please don't duplicate reorder kernels

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp (hunk 4 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp]: done

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp (hunk 2 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/jitter.cpp]: done

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data.cl [hunk 6]
```diff
@@ -132,7 +132,11 @@ KERNEL (reorder_data)(
     float Ycomponent = mad(Y.x, 296.82f, -18.624f);
     float res = clamp(Ycomponent, 0.f, 255.f);
 #else
-    CALC_TYPE res = TO_CALC_TYPE(input[input_idx]);
+    #ifdef BF16_INPUT
+        CALC_TYPE res = _convert_as_bfloat16_float(input[input_idx]);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data.cl:136]: isn't it necessary to have additional macro to convert type to CALC_TYPE?
I mean, `CALC_TYPE res = TO_CALC_TYPE(_convert_as_bfloat16_float(input[input_idx]));`

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp [hunk 7]
```diff
@@ -1489,6 +1489,16 @@ JitConstants MakeTypeJitConstants(Datatype dataType, const std::string& macroNam
             type_size = "0.5f";
             is_fp = false;
             break;
+        case Datatype::BF16:
+            type = "ushort";
+            val_one = "(ushort) 1";
+            val_zero = "(ushort) 0";
+            to_type = "_convert_bfloat16_as_ushort(v)";
+            to_type_sat = "_convert_bfloat16_as_ushort(v)";
+            as_type = "_convert_bfloat16_as_ushort(v)";
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/jitter.cpp:1498]: behavior of as_type is to keep the current bits as-is. But it is changing bits. I think it is not an expected behavior..

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_ref.cpp [hunk 8]
```diff
@@ -100,6 +100,7 @@ static std::string GetReorderedOutputOrder(const permute_params& params, const s
 
 ParamsKey PermuteKernelRef::GetSupportedKey() const {
     ParamsKey k;
+    k.EnableInputDataType(Datatype::BF16);
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_ref.cpp:103]: does permute support bf16 input? I don't see any change from permute cl kernel..

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        std::cout << "znaleziono " << n->get_friendly_name() << std::endl;
+        if ( n->get_friendly_name() == "conversion_bf16_to_f32" ) {
+            ASSERT_EQ(n->get_input_element_type(0),  ov::element::bf16);
+            ASSERT_EQ(n->get_output_element_type(0),  ov::element::f32);
+            convert_found = true;
+        }
+    }
+    ASSERT_TRUE(convert_found);
+}
+*/
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:85]: is it a necessary test case?

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    float given;
+    float expected;
+    std::tie(given, expected) = GetParam();
+    std::shared_ptr<ov::Model> function_without_convert = init_subgraph_without_convert(given);
+    ov::CompiledModel model_without_convert = core->compile_model(function_without_convert, targetDevice, {ov::hint::inference_precision(ov::element::f32)});
+    auto request = model_without_convert.create_infer_request();
+    request.infer();
+    auto out_tensor = request.get_output_tensor();
+    for ( size_t i = 0 ; i < out_tensor.get_size(); i++ ) {
+        ASSERT_NEAR(reinterpret_cast<unsigned short*>(out_tensor.data())[i], expected, 0.2f); //3072
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:118]: it is difficult to understand the meaning of this comment. `3072` could you explain more?

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+INSTANTIATE_TEST_SUITE_P(Inference_without_convert, BF16WeightsDecompression,
+                         testing::Values(BF16WeightsDecompressionParams(768, 17728.0f),
+                            BF16WeightsDecompressionParams(3.0f, 16704.0f),
+                            BF16WeightsDecompressionParams(2.0f, 16640.0f),
+                            BF16WeightsDecompressionParams(49512.0f, 18497.0f), //768x64
+                            BF16WeightsDecompressionParams(98304.0f, 18624.0f), //768x128
+                            BF16WeightsDecompressionParams(55340232221128654848.0f, 24896.0f),
+                            BF16WeightsDecompressionParams(-55340232221128654848.0f, 57664.0f),
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:129]: are these numbers generated by reinterpreting bf16 values? Then what about having an explicit function to convert/reinterpreting function?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_ref.cpp (hunk 8 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_ref.cpp:103]: it is needed as permute is done to transpose matrix before multiply
image

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp (hunk 9 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:85]: no

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp (hunk 10 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:118]: sorry it is here by mistake, I will delete it

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp (hunk 11 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:129]: I don't think It is good idea to check function call by another function call

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data.cl (hunk 6 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/reorder_data.cl]: done, added macro

File: src/plugins/intel_gpu/src/kernel_selector/jitter.cpp (hunk 7 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/src/kernel_selector/jitter.cpp]: deleted as_type definition at the moment

File: src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp (hunk 11 above)
Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:112]: but I found that way to check in framework, what should I do?

Comment (isanghao) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:112]: the problem with current code is that it is difficult to understand the meaning of these numbers. Then what about adding more comments about these numbers?

Comment (michal-miotk) [reply src/plugins/intel_gpu/tests/functional/subgraph_tests/dynamic/convert_bf16_fp32_fp16.cpp:112]: I think that I will add one function, then no long description needed

**Review by p-durandin — APPROVED** [review]

**Review by isanghao — APPROVED** [review]

**Pull request merged**